### PR TITLE
R spec users

### PR DESCRIPTION
**Pull request by mckhannah1901**
### JIRA
[ATH-1143_User_RSpec_Tests](https://wpengine.atlassian.net/secure/RapidBoard.jspa?rapidView=522&projectKey=ATH&modal=detail&selectedIssue=ATH-1143)

### What was done
- Added RSpec testing to user controller and model

### How to test
- Run bundle exec rspec

### CODE REVIEW

**Comment by mckhannah1901** [comment]
Any suggestions about more tests I can do or better ways of testing are welcome!

File: landing_page/spec/controllers/user_controller_spec.rb [hunk 1]
```diff
@@ -0,0 +1,27 @@
+require 'rails_helper'
+
+RSpec.describe UsersController, type: :controller do
+
+    context "create a new user account with acceptable credentials" do
```
Comment (sbodwyer) [inline landing_page/spec/controllers/user_controller_spec.rb]: Could add tests for other controller methods too like update show etc

Comment (mckhannah1901) [reply landing_page/spec/controllers/user_controller_spec.rb]: Cool, I'll get onto that now 👍

Comment (sbodwyer) [reply landing_page/spec/controllers/user_controller_spec.rb]: When testing a unit of code its useful to look at what the public methods are, as each warrant at least one unit test

**Comment by mckhannah1901** [comment]
I'm using the command 'rake stats'

File: landing_page/Gemfile.lock [hunk 2]
```diff
@@ -78,8 +78,10 @@ GEM
       railties (>= 4.1.0)
       responders
       warden (~> 1.2.3)
+    diff-lcs (1.3)
```
Comment (alvarocavalcanti) [inline landing_page/Gemfile.lock]: When trying to install the dependencies with `bundle install` I got this error:

```bash
Bundler could not find compatible versions for gem "method_source":
  In snapshot (Gemfile.lock):
    method_source (= 1.0.0)

  In Gemfile:
    pry-nav was resolved to 0.3.0, which depends on
      pry (>= 0.9.10, < 0.13.0) was resolved to 0.12.2, which depends on
        method_source (~> 0.9.0)

    devise was resolved to 4.7.1, which depends on
      responders was resolved to 3.0.0, which depends on
        railties (>= 5.0) was resolved to 6.0.2.2, which depends on
          method_source

Running `bundle update` will rebuild your snapshot from scratch, using only
the gems in your Gemfile, which may resolve the conflict.
```

Running `bundle update` fixed it, but left me with a modified `Gemfile.lock`. It seems that the Gemfile ciuld use some tidying up. Consider this, if I have not done any code change, just pulled in the latest from the repo, there should be no changes to any file just by installing the dependencies.

File: landing_page/spec/controllers/user_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+        it "successfully adds a new user" do
+            expect{create_new_user}.to change(User, :count).by(1)
+        end
+
+        it "logs the user in upon account creation" do
+            session_id  = session[:user_id]
```
Comment (alvarocavalcanti) [inline landing_page/spec/controllers/user_controller_spec.rb]: This is confusing. How is it there's a `session[:user_id]`?

File: landing_page/spec/controllers/user_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+            create_new_user
+            expect(session[:user_id]).to_not eq(session_id)
```
Comment (alvarocavalcanti) [inline landing_page/spec/controllers/user_controller_spec.rb]:
```suggestion
            expect(session[:user_id]).to_not be_nil
```

File: landing_page/spec/controllers/user_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+        end
+
+        it "redirects a new user to their profile page upon creation" do
+            create_new_user
+            expect(response).to redirect_to(assigns(:user))
+        end
+
+        def create_new_user
```
Comment (alvarocavalcanti) [inline landing_page/spec/controllers/user_controller_spec.rb]: Ideally, this method should be private.

File: landing_page/spec/models/user_spec.rb [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        end
+
+        it "fails if surname is longer than 50 letters" do
+            user.second_name = "abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyz"
+            expect(user).to_not be_valid
+        end
+
+    end
+    
+end
```
Comment (alvarocavalcanti) [inline landing_page/spec/models/user_spec.rb]: Missing empty line at the end of file.

File: landing_page/spec/models/user_spec.rb [hunk 7]
```diff
@@ -0,0 +1,90 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+
+    let (:user){User.new}
```
Comment (alvarocavalcanti) [inline landing_page/spec/models/user_spec.rb]: Why do you defined this variable only to re-define it in the context below? This is unnecessary and can be removed.

File: landing_page/spec/models/user_spec.rb [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+    context "don't create a user account with unacceptable credentials" do
```
Comment (alvarocavalcanti) [inline landing_page/spec/models/user_spec.rb]: The context should represent a "state", a "pre-condition", not a behaviour, which would fall to the "it" blocks. Thus, I suggest the following:

```suggestion
    context "user account with unacceptable data" do
```

File: landing_page/spec/models/user_spec.rb [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+
+        it "fails if the email does not match the required format - 3" do
+            user.email = "1@gmail"
+            expect(user).to_not be_valid
+        end
+
+        it "fails if the email does not match the required format - 4" do
+            user.email = "1"
+            expect(user).to_not be_valid
+        end
```
Comment (alvarocavalcanti) [inline landing_page/spec/models/user_spec.rb]: These could be merged into a single block with more than one assertion.

Either this way:
```suggestion
        it "fails if the email is invalid" do
            user.email = "1.com"
            expect(user).to_not be_valid
            
            user.email = "@gmail.com"
            expect(user).to_not be_valid
            
            user.email = "1@gmail"
            expect(user).to_not be_valid
            
            user.email = "1"
            expect(user).to_not be_valid
        end
```

Or this way:
```suggestion
        it "fails if the email is invalid" do
            ["1.com", "@gmail.com", "1@gmail", "1"].each do |email|
                user.email = email
                expect(user).to_not be_valid
            end            
        end
```

Which one do you prefer?

File: landing_page/spec/models/user_spec.rb [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+
+        it "fails if the email has already been taken up by another user" do
+            User.new(:first_name => 'Testing', :second_name => 'Tested', :email => 'testing@tested.test',
+                    :password => 'password', :password_confirmation => 'password').save
+            user2 = User.new(:first_name => 'Second', :second_name => 'Test', :email => 'testing@tested.test',
```
Comment (alvarocavalcanti) [inline landing_page/spec/models/user_spec.rb]: This variable name is misleading. This is **not** a `User` object, it's a boolean returned from the `save` method.

```suggestion
            user_created = User.new(:first_name => 'Second', :second_name => 'Test', :email => 'testing@tested.test',
```

**Comment by alvarocavalcanti** [comment]
> It may not be part of this PR but how do you propose to measure the coverage of your unit tests?

The parent issue, https://wpengine.atlassian.net/browse/ATH-1142, has been updated with the coverage information, as well as how to verify it.

But this PR should have the instructions on how to get that stats.

File: landing_page/Gemfile.lock (hunk 2 above)
Comment (mckhannah1901) [reply landing_page/Gemfile.lock]: Resolved in commit d8d58c516f262cbd675e9faa70f4fbc4bfafee01

File: landing_page/spec/controllers/user_controller_spec.rb (hunk 4 above)
Comment (mckhannah1901) [reply landing_page/spec/controllers/user_controller_spec.rb]: Resolved in commit 0a7fed3763eb9f6d8481ed307599325f2e63d1cd

File: landing_page/spec/controllers/user_controller_spec.rb (hunk 3 above)
Comment (mckhannah1901) [reply landing_page/spec/controllers/user_controller_spec.rb]: Resolved in commit 0a7fed3763eb9f6d8481ed307599325f2e63d1cd

File: landing_page/Gemfile [hunk 11]
```diff
@@ -50,9 +52,12 @@ group :development do
   # Spring speeds up development by keeping your application running in the background. Read more: https://github.com/rails/spring
   gem 'spring'
   gem 'spring-watcher-listen', '~> 2.0.0'
+  gem 'rails-controller-testing'
```
Comment (jamesfairbairn-wpengine) [inline landing_page/Gemfile]: Just wondering, do you need to add this gem here as well? I may be wrong, but I think that adding it to the `group :development :test` above means that it's already there when you install develop.

File: landing_page/Gemfile [hunk 12]
```diff
@@ -50,9 +52,12 @@ group :development do
...
 end
 
 # Windows does not include zoneinfo files, so bundle the tzinfo-data gem
 gem 'tzinfo-data', platforms: [:mingw, :mswin, :x64_mingw, :jruby]
 
+gem 'pry'
```
Comment (jamesfairbairn-wpengine) [inline landing_page/Gemfile]: These may be better put within either the `:development` or `:test` groups, as you wouldn't really need these in production.

File: landing_page/Gemfile (hunk 11 above)
Comment (mckhannah1901) [reply landing_page/Gemfile]: Good catch, I'll fix that now

File: landing_page/Gemfile (hunk 12 above)
Comment (mckhannah1901) [reply landing_page/Gemfile]: Resolved in commit 07d7e2f4dd05baad296d964d81b1d24900b8a7d0

File: landing_page/Gemfile (hunk 11 above)
Comment (mckhannah1901) [reply landing_page/Gemfile]: Resolved in commit 07d7e2f4dd05baad296d964d81b1d24900b8a7d0

File: landing_page/spec/controllers/user_controller_spec.rb (hunk 5 above)
Comment (mckhannah1901) [reply landing_page/spec/controllers/user_controller_spec.rb]: Resolved in commit ec7fcdd07197a2e7f1ac603765cff90c46f846f4

File: landing_page/spec/models/user_spec.rb (hunk 7 above)
Comment (mckhannah1901) [reply landing_page/spec/models/user_spec.rb]: Resolved in commit 2b36cd84305e63b4adadc6c33091e49dfe3589ef

File: landing_page/spec/models/user_spec.rb (hunk 6 above)
Comment (mckhannah1901) [reply landing_page/spec/models/user_spec.rb]: commit 2f95811e86b77e832ca3bb88c97385ba5c360ff9

File: landing_page/spec/models/user_spec.rb (hunk 8 above)
Comment (mckhannah1901) [reply landing_page/spec/models/user_spec.rb]: commit a8b0890076739c912f85351d691d88a8f5d1ca6a

File: landing_page/spec/models/user_spec.rb (hunk 9 above)
Comment (mckhannah1901) [reply landing_page/spec/models/user_spec.rb]: I preferred the first suggestion, I just feel like it's easier to read 👍 

Resolved in commit 302892f21532d473cba8b1747c5b20e84a23bf73

File: landing_page/spec/models/user_spec.rb (hunk 10 above)
Comment (mckhannah1901) [reply landing_page/spec/models/user_spec.rb]: Resolved in commit d28b3732553795140f5a159faae9d35e9eb4e2a6

File: landing_page/spec/controllers/user_controller_spec.rb [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+        it "redirects a new user to their profile page upon creation" do
+            create_new_user
+            expect(response).to redirect_to(assigns(:user))
+        end
+
+    end
+
+    context "GET #index" do
+
+        it "shows the list of users, expects a redirect response to be 200" do
```
Comment (alvarocavalcanti) [inline landing_page/spec/controllers/user_controller_spec.rb]: Ideally, test scenarios should not have expectations as part of their names. A scenario should describe what should happen, "show the list of users", but not how that verification is made. The point here is that the method of verification could change and you would have to update the scenario name just because of it.

And in this specific case, what is the evidence that the `#index` shows the list of the users? Just a `200` response? Is that enough? If I have 10 users and my list only shows 5, but returns a `200` status is that correct?

File: landing_page/spec/controllers/user_controller_spec.rb [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+
+    context "GET #show" do
+
+        it "shows a user their profile, expect redirect response to be 200" do
+            user = FactoryBot.create :user
+            user.save
+            login(user)
+
+            get :show, params: { id: user.id }
+            expect(response.status).to eq(200)
```
Comment (alvarocavalcanti) [inline landing_page/spec/controllers/user_controller_spec.rb]: Same as my previous comment. What if the page is blank but it returns a `200`?

File: landing_page/spec/controllers/user_controller_spec.rb [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+
+            expect{delete :destroy, params: { id: user.id }}.to change(User, :count).by(-1)
+        end
+
+    end
+
+    context "PUT #update" do
+
+        it "allows a user to update their data" do
+            user = FactoryBot.create :user
```
Comment (alvarocavalcanti) [inline landing_page/spec/controllers/user_controller_spec.rb]: Although the expectation is correct, checking the name that was provided in the `:update` call, it's a good practice to make the setup more explicit. Something along the lines of:

```suggestion
            user = FactoryBot.create :user, name: "Not Jane"
```

That way it's clear where the reader should focus on.

File: landing_page/spec/controllers/user_controller_spec.rb (hunk 14 above)
Comment (mckhannah1901) [reply landing_page/spec/controllers/user_controller_spec.rb]: The page shouldn't be blank. In order to access this page, a user needs to be logged in, so their data will have been saved  and this is what they're accessing. I'll look into a better way of verifying this than just a 200 response though

File: landing_page/spec/controllers/user_controller_spec.rb (hunk 15 above)
Comment (mckhannah1901) [reply landing_page/spec/controllers/user_controller_spec.rb]: Resolved in commit commit d97be7bed27ec18d0421cdb237a47d222374017a

File: landing_page/spec/controllers/user_controller_spec.rb [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+        it "shows the list of users, expects a redirect response to be 200" do
+            get :index
+            expect(response.status).to eq(200)
+        end
+
+    end
+
+    context "GET #show" do
+
+        it "shows a user their profile, expect redirect response to be 200" do
```
Comment (alvarocavalcanti) [inline landing_page/spec/controllers/user_controller_spec.rb]: The scenario description should not have the expectation.

```suggestion
        it "shows a user their profile" do
```

Additionally, this expectation part is wrong, this scenario does not involve a redirect. With that in mind, a couple os questions:

1. What do you mean with "redirect response"?
1. What does the `200` mean? What is it?

File: landing_page/spec/controllers/user_controller_spec.rb (hunk 14 above)
Comment (alvarocavalcanti) [reply landing_page/spec/controllers/user_controller_spec.rb]: Consider the proper scenario, "shows a user their profile", how can we assert that the provided user is seeing their profile, and not another user's?

Also, you wrote:
> The page shouldn't be blank.

That's your assumption based on the code, or in other words, that's the code intention. The role of a test is to challenge that assumption by explicitly testing the resulting data.

File: landing_page/spec/controllers/user_controller_spec.rb (hunk 13 above)
Comment (mckhannah1901) [reply landing_page/spec/controllers/user_controller_spec.rb]: I have added further tests for the index. The 200 response isn't adequate on it's own.

Please see commit fa0508cf969ee02dc88928e9c412d18df6ea8a74 for updated tests

File: landing_page/spec/controllers/user_controller_spec.rb (hunk 14 above)
Comment (mckhannah1901) [reply landing_page/spec/controllers/user_controller_spec.rb]: Resolved in commit fa0508cf969ee02dc88928e9c412d18df6ea8a74

File: landing_page/spec/controllers/user_controller_spec.rb (hunk 16 above)
Comment (mckhannah1901) [reply landing_page/spec/controllers/user_controller_spec.rb]: They overlap. So the 200 is to show that the redirect response was a success. This has since been altered in commit fa0508cf969ee02dc88928e9c412d18df6ea8a74

File: landing_page/Gemfile [hunk 17]
```diff
@@ -50,6 +53,8 @@ group :development do
   # Spring speeds up development by keeping your application running in the background. Read more: https://github.com/rails/spring
   gem 'spring'
   gem 'spring-watcher-listen', '~> 2.0.0'
+  gem 'pry'
```
Comment (sbodwyer) [inline landing_page/Gemfile]: would recommend moving the pry gems to test, development as it is sometimes useful to breakpoint during a test

File: landing_page/spec/controllers/user_controller_spec.rb [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+            get :index
+            expect(response.status).to eq(200)
+        end
+
+    end
+
+    context "GET #show" do
+
+        it "shows a user their profile" do
+            user = FactoryBot.create :user
```
Comment (sbodwyer) [inline landing_page/spec/controllers/user_controller_spec.rb]: Rather than create a new user with the factory here you can use the same user object if you place the  factory creation on line 25 to the parent context (after line 5). If this is confusing gimme a call to discuss

File: landing_page/spec/controllers/user_controller_spec.rb [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+
+            expect{delete :destroy, params: { id: user.id }}.to change(User, :count).by(-1)
+        end
+
+    end
+
+    context "PUT #update" do
+
+        it "allows a user to update their data" do
+            user = FactoryBot.create :user, first_name: "Not Jane"
```
Comment (sbodwyer) [inline landing_page/spec/controllers/user_controller_spec.rb]: same as previous comment on factory creations

File: landing_page/Gemfile (hunk 17 above)
Comment (mckhannah1901) [reply landing_page/Gemfile]: That's a good catch, thanks for spotting it!

File: landing_page/spec/controllers/user_controller_spec.rb [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+        it "shows a user their profile" do
+            user = FactoryBot.create :user
+            user.save
+            login(user)
+
+            get :show, params: { id: user.id }
+            expect(response).to render_template('show')
+            expect(response.status).to eq(200)
+        end
+
```
Comment (sbodwyer) [inline landing_page/spec/controllers/user_controller_spec.rb]: Can you think of another test here on the show controller that might be relevant? What if the user is not logged in?

File: landing_page/spec/controllers/user_controller_spec.rb [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+    end
+
+    context "DELETE #destroy" do
+
+        it "deletes a user from the database" do
+            user = FactoryBot.create :user
```
Comment (sbodwyer) [inline landing_page/spec/controllers/user_controller_spec.rb]: same as previous factory creation comment

File: landing_page/spec/controllers/user_controller_spec.rb [hunk 22]
```diff
@@ -0,0 +1,100 @@
...
+            user.save
+            login(user)
+
+            expect{delete :destroy, params: { id: user.id }}.to change(User, :count).by(-1)
+        end
+
```
Comment (sbodwyer) [inline landing_page/spec/controllers/user_controller_spec.rb]: Is there another test you can think of here?

File: landing_page/spec/controllers/user_controller_spec.rb [hunk 23]
```diff
@@ -0,0 +1,100 @@
...
+    context "PUT #update" do
+
+        it "allows a user to update their data" do
+            user = FactoryBot.create :user, first_name: "Not Jane"
+            user.save
+            login(user)
+
+            put :update, params: { :id => user.id, :user => { :first_name => "Jane", :second_name => "Doe",
+                                :email => "factorybot@test.com", :password => "password", :password_confirmation => "password"}}
```
Comment (sbodwyer) [inline landing_page/spec/controllers/user_controller_spec.rb]: is specifying all these params necessary? I believe you can just specify the first name and nothing else. If you only assert on the first name that is all that is required

File: landing_page/spec/controllers/user_controller_spec.rb [hunk 24]
```diff
@@ -0,0 +1,100 @@
...
+            user.reload
+
+            expect(user.first_name).to eq("Jane")
+            expect(response).to redirect_to(assigns(:user))
+        end
+
```
Comment (sbodwyer) [inline landing_page/spec/controllers/user_controller_spec.rb]: Can you think of another test? What is a user is not logged in and they try to update params?

File: landing_page/spec/controllers/user_controller_spec.rb (hunk 18 above)
Comment (mckhannah1901) [reply landing_page/spec/controllers/user_controller_spec.rb]: I didn't put it to that line, but created a new context before any tests that require FactoryBot. So it's still only created once, like you suggested 👍 

Resolved in commit 27ae0aec5f07e47a8a8f6aeb06f6f1939103444f

File: landing_page/spec/controllers/user_controller_spec.rb (hunk 21 above)
Comment (mckhannah1901) [reply landing_page/spec/controllers/user_controller_spec.rb]: Resolved in commit 27ae0aec5f07e47a8a8f6aeb06f6f1939103444f

File: landing_page/spec/controllers/user_controller_spec.rb (hunk 19 above)
Comment (mckhannah1901) [reply landing_page/spec/controllers/user_controller_spec.rb]: Resolved in commit 27ae0aec5f07e47a8a8f6aeb06f6f1939103444f

File: landing_page/spec/controllers/user_controller_spec.rb (hunk 23 above)
Comment (mckhannah1901) [reply landing_page/spec/controllers/user_controller_spec.rb]: I tried doing this but it fails as  it doesn't think the first name has been updated so maybe the params are necessary. I'm not too sure but I'll look into it

File: landing_page/spec/controllers/user_controller_spec.rb (hunk 24 above)
Comment (mckhannah1901) [reply landing_page/spec/controllers/user_controller_spec.rb]: Really good suggestion with this as it showed access levels to urls I wasn't aware of. I've now implemented it in commit b6c85798251731e784dea32679ecb2dfca592286 😄

File: landing_page/spec/controllers/user_controller_spec.rb (hunk 20 above)
Comment (mckhannah1901) [reply landing_page/spec/controllers/user_controller_spec.rb]: Resolved in commit b6c85798251731e784dea32679ecb2dfca592286

File: landing_page/spec/controllers/user_controller_spec.rb (hunk 22 above)
Comment (mckhannah1901) [reply landing_page/spec/controllers/user_controller_spec.rb]: Yup I added an additional test in commit 03286eddfd018d0b58079cbfc4ac2146f8971b26 to test what happens if you try to delete a user when not logged in. Also added an extra expectation for the original delete test in commit 829c95d747b3876ba53a903c172196ad58852f78

File: landing_page/spec/controllers/user_controller_spec.rb (hunk 23 above)
Comment (mckhannah1901) [reply landing_page/spec/controllers/user_controller_spec.rb]: I think the params are necessary in this case as it's validated by the presence of all the params

File: landing_page/app/models/ability.rb [hunk 25]
```diff
@@ -4,9 +4,16 @@ class Ability
...
+       user.present? && user.id == u.id
     end
 
+    can :update, User do |u|
+      user.present? && user.id == u.id
+    end
+
+    can :destroy, User do |u|
+      user.present? && user.id == u.id
+    end
```
Comment (alvarocavalcanti) [inline landing_page/app/models/ability.rb]: How can we remove this duplication?

File: landing_page/spec/controllers/user_controller_spec.rb [hunk 26]
```diff
@@ -0,0 +1,124 @@
...
+    end
+
+    context "FactoryBot tests" do
+
+    before :each do
+        @user = FactoryBot.create(:user)
+        session[:user_id] = @user.id
+        login(@user)
+    end
```
Comment (alvarocavalcanti) [inline landing_page/spec/controllers/user_controller_spec.rb]: Wrong indentation.

File: landing_page/spec/controllers/user_controller_spec.rb [hunk 27]
```diff
@@ -0,0 +1,124 @@
...
+
+        context "GET #index" do
+
+            it "assigns users" do
+                get :index
+                expect(assigns(:users)).to eq([@user])
+            end
```
Comment (alvarocavalcanti) [inline landing_page/spec/controllers/user_controller_spec.rb]: This is correct! Although I'd rather we keep the previous scenario description "lists the users".

File: landing_page/spec/controllers/user_controller_spec.rb [hunk 28]
```diff
@@ -0,0 +1,124 @@
...
+
+            it "renders the template for the index" do
+                get :index
+                expect(response).to render_template('index')
+            end
```
Comment (alvarocavalcanti) [inline landing_page/spec/controllers/user_controller_spec.rb]: This scenario should be removed as it solely tests out Rails code. You see, you are only testing if the `index` template is rendered upon sending a get request to the `index` action of the controller.

File: landing_page/spec/controllers/user_controller_spec.rb [hunk 29]
```diff
@@ -0,0 +1,124 @@
...
+
+            it "returns a succesful response" do
+                get :index
+                expect(response.status).to eq(200)
+            end
```
Comment (alvarocavalcanti) [inline landing_page/spec/controllers/user_controller_spec.rb]: This one could live separately as-is but I would suggest it being merged to the first scenario. My reasoning is, it has the same setup as the the other scenario and actually it's part of it, which means that we're sending the exact same request. 

If we merge them we would still be able to identify individually what failed based on the line number and that would not increase the test readability nor complexity, which would be valid points for having two separate scenarios.

File: landing_page/spec/controllers/user_controller_spec.rb [hunk 30]
```diff
@@ -0,0 +1,124 @@
...
+
+        end
+
+        context "GET #show" do
+
+            it "shows a user their profile" do
+                get :show, params: { id: @user.id }
+                expect(response).to render_template('show')
+                expect(response.status).to eq(200)
+            end
```
Comment (alvarocavalcanti) [inline landing_page/spec/controllers/user_controller_spec.rb]: We need to test the actual content of the response, code coverage alone isn't enough, we need to write valuable tests.

I modified the controller, commenting out this line `# @user = User.find(params[:id])` and the test still passes, and that's bad.

File: landing_page/spec/controllers/user_controller_spec.rb [hunk 31]
```diff
@@ -0,0 +1,124 @@
...
+                get :index
+                expect(assigns(:users)).to eq([@user])
+            end
+
+            it "renders the template for the index" do
+                get :index
+                expect(response).to render_template('index')
+            end
+
+            it "returns a succesful response" do
```
Comment (sbodwyer) [inline landing_page/spec/controllers/user_controller_spec.rb]: mis-spelling

Comment (mckhannah1901) [reply landing_page/spec/controllers/user_controller_spec.rb]: Resolved in commit a915a552e396c7fb923192241e063af579233b79

File: landing_page/spec/controllers/user_controller_spec.rb (hunk 27 above)
Comment (mckhannah1901) [reply landing_page/spec/controllers/user_controller_spec.rb]: Resolved in commit b4aadfca7384abfd436b5d5c156102b211b88707

File: landing_page/spec/controllers/user_controller_spec.rb (hunk 26 above)
Comment (mckhannah1901) [reply landing_page/spec/controllers/user_controller_spec.rb]: Resolved in commit b4aadfca7384abfd436b5d5c156102b211b88707

File: landing_page/spec/controllers/user_controller_spec.rb (hunk 28 above)
Comment (mckhannah1901) [reply landing_page/spec/controllers/user_controller_spec.rb]: Resolved in commit 4871a5281315b569ef730a49585feaab4de96af4

File: landing_page/spec/controllers/user_controller_spec.rb (hunk 29 above)
Comment (mckhannah1901) [reply landing_page/spec/controllers/user_controller_spec.rb]: Resolved in commit 4871a5281315b569ef730a49585feaab4de96af4

File: landing_page/spec/controllers/user_controller_spec.rb [hunk 32]
```diff
@@ -0,0 +1,100 @@
...
+
+        it "renders the template for the index" do
+            get :index
+            expect(response).to render_template('index')
+        end
+
+        it "returns a succesful response" do
+            get :index
+            expect(response.status).to eq(200)
+        end
```
[the remark this answers is not in the document]
Comment (mckhannah1901) [reply landing_page/spec/controllers/user_controller_spec.rb]: Resolved in commit 4871a5281315b569ef730a49585feaab4de96af4

File: landing_page/spec/controllers/user_controller_spec.rb (hunk 30 above)
Comment (mckhannah1901) [reply landing_page/spec/controllers/user_controller_spec.rb]: commit 804d4cb956b489ae0d19a97f91cb1949ba834576

File: landing_page/spec/controllers/user_controller_spec.rb [hunk 33]
```diff
@@ -0,0 +1,124 @@
...
+
+        it "redirects a new user to their profile page upon creation" do
+            create_new_user
+            expect(response).to redirect_to(assigns(:user))
+        end
+
+    end
+
+    context "FactoryBot tests" do
```
Comment (sbodwyer) [inline landing_page/spec/controllers/user_controller_spec.rb]: Context descriptions should be about the stuff you are testing not about the testing / mocking frameworks. You could delete this context and move the factory creation to line 4 after `RSpec.describe UsersController ...`

Comment (mckhannah1901) [reply landing_page/spec/controllers/user_controller_spec.rb]: When I did that, it caused all of my tests to fail, including the first three that don't use it as it conflicted with the create_new_user method. I can change the description of the context from "FactoryBots test" to something more appropriate

Comment (sbodwyer) [reply landing_page/spec/controllers/user_controller_spec.rb]: ok one sec , I'll see if I can fix

Comment (mckhannah1901) [reply landing_page/spec/controllers/user_controller_spec.rb]: Resolved in commit 5c9ac03bcfe8b27ff22e3b5a6fd61a93633dbafe

File: landing_page/app/controllers/users_controller.rb [hunk 34]
```diff
@@ -60,9 +43,9 @@ def update
     end
   end
 
-  # DELETE /users/1
-  # DELETE /users/1.json
   def destroy
+    authorize! :update, @user, :message => "Unable to update user"
```
Comment (sbodwyer) [inline landing_page/app/controllers/users_controller.rb]: Message should be "unable to destroy user"?

File: landing_page/spec/factories/users.rb [hunk 35]
```diff
@@ -0,0 +1,10 @@
+FactoryBot.define do
+    factory :user do
+      first_name { "John" }
```
Comment (sbodwyer) [inline landing_page/spec/factories/users.rb]: Have you heard of the faker gem? Consider if a test was to create two users via the factory, they both would have the same data at the moment. Faker will generate unique fields for your data so that this doesn't happen. Maybe this change can be made in a separate PR , what do you think @alvarocavalcanti ?

File: landing_page/spec/views/users/show.html.erb_spec.rb [hunk 36]
```diff
@@ -0,0 +1,35 @@
+require 'rails_helper'
+
+describe "users/show.html.erb" do
+
+    context "when the user has data" do 
+
+        it "displays a users profile correctly" do
+
```
Comment (sbodwyer) [inline landing_page/spec/views/users/show.html.erb_spec.rb]: can remove this blank line

File: landing_page/spec/factories/users.rb (hunk 35 above)
Comment (mckhannah1901) [reply landing_page/spec/factories/users.rb]: Yeah I've heard of the Faker gem, I've used it before when I was playing around with creating a blog post and it was quite interesting to use

File: landing_page/spec/views/users/show.html.erb_spec.rb [hunk 37]
```diff
@@ -0,0 +1,35 @@
...
+            expect(rendered).to have_content("test@meh.com")
+
+        end
+
+    end
+
+    context "when the user has no data" do
+
+        it "doesn't display content" do
+
```
Comment (sbodwyer) [inline landing_page/spec/views/users/show.html.erb_spec.rb]: remove blank line

File: landing_page/spec/views/users/show.html.erb_spec.rb [hunk 38]
```diff
@@ -0,0 +1,35 @@
...
+            render
+
+            expect(rendered).to have_content("Test")
+            expect(rendered).to have_content("Meh")
+            expect(rendered).to have_content("test@meh.com")
+
+        end
+
+    end
```
Comment (sbodwyer) [inline landing_page/spec/views/users/show.html.erb_spec.rb]: blank line

File: landing_page/spec/views/users/show.html.erb_spec.rb [hunk 39]
```diff
@@ -0,0 +1,35 @@
...
+
+            render
+
+            expect(rendered).to have_content(nil)
+
+        end
+
+    end
+
+end
```
Comment (sbodwyer) [inline landing_page/spec/views/users/show.html.erb_spec.rb]: more blanks - install rubocop?

Comment (mckhannah1901) [reply landing_page/spec/views/users/show.html.erb_spec.rb]: Rubocop is on the next story for me after the current one but I'll remove the lines now 👍

File: landing_page/spec/views/users/show.html.erb_spec.rb (hunk 37 above)
Comment (mckhannah1901) [reply landing_page/spec/views/users/show.html.erb_spec.rb]: Resolved in commit bcc1c2e3f03b36170702a98979328c0ce1cd1961

File: landing_page/spec/views/users/show.html.erb_spec.rb (hunk 39 above)
Comment (mckhannah1901) [reply landing_page/spec/views/users/show.html.erb_spec.rb]: Resolved in commit bcc1c2e3f03b36170702a98979328c0ce1cd1961

File: landing_page/spec/views/users/show.html.erb_spec.rb (hunk 38 above)
Comment (mckhannah1901) [reply landing_page/spec/views/users/show.html.erb_spec.rb]: Resolved in commit bcc1c2e3f03b36170702a98979328c0ce1cd1961

File: landing_page/spec/views/users/show.html.erb_spec.rb (hunk 36 above)
Comment (mckhannah1901) [reply landing_page/spec/views/users/show.html.erb_spec.rb]: Resolved in commit bcc1c2e3f03b36170702a98979328c0ce1cd1961

File: landing_page/app/controllers/users_controller.rb (hunk 34 above)
Comment (mckhannah1901) [reply landing_page/app/controllers/users_controller.rb]: Resolved in commit 9bd8da11cba94945e7262cefa9a143ad3b3d93b7

File: landing_page/spec/factories/users.rb (hunk 35 above)
Comment (alvarocavalcanti) [reply landing_page/spec/factories/users.rb]: Although I like the usage of `Faker` maybe we don't have a use case for it for the moment.

But a scenario in which using it would be required would be to render a list of articles or users, at least two, and assert that they are unique.

File: landing_page/app/models/ability.rb (hunk 25 above)
Comment (jamesfairbairn-wpengine) [reply landing_page/app/models/ability.rb]: I think you might be able to do something like:
```ruby
can [:view, :update, :destroy], User do |u|
    user.present? && user.id == u.id
end
```
